lib/spack/llnl/util/tty/log.py

```diff
@@ -593,25 +658,37 @@ def _writer_daemon(stdin, read_fd, write_fd, echo, log_file, control_pipe):
 
     Arguments:
         stdin (stream): input from the terminal
-        read_fd (int): pipe for reading from parent's redirected stdout
-        write_fd (int): parent's end of the pipe will write to (will be
-            immediately closed by the writer daemon)
+        read (int): pipe for reading from parent's redirected stdout
         echo (bool): initial echo setting -- controlled by user and
             preserved across multiple writer daemons
         log_file (file-like): file to log all output
         control_pipe (Pipe): multiprocessing pipe on which to send control
             information to the parent
 
     """
+    # If this process was forked, then it will inherit file descriptors from
+    # the parent process. This process depends on closing all instances of
+    # write_fd to terminate the reading loop, so we close the file descriptor
+    # here. Forking is the process spawning method everywhere except Mac OS
+    # for Python >= 3.8
+    if sys.version_info < (3, 8) or sys.platform != 'darwin':
+        os.close(write_fd)
+
     # Use line buffering (3rd param = 1) since Python 3 has a bug
     # that prevents unbuffered text I/O.
-    in_pipe = os.fdopen(read_fd, 'r', 1)
-    os.close(write_fd)
+    in_pipe = os.fdopen(read_wrapper._handle, 'r', 1)
+
+    if stdin_wrapper:
+        stdin = os.fdopen(stdin_wrapper._handle)
+    else:
+        stdin = None
 
     # list of streams to select from
     istreams = [in_pipe, stdin] if stdin else [in_pipe]
     force_echo = False      # parent can force echo for certain output
 
+    log_file = log_file_wrapper.unwrap()
```

Comment:
> This seems fragile. We're taking a class/method that can return a stream or a filename, and then we're assuming it's a stream and writing to it.

`FileWrapper.unwrap` is always supposed to return a file-like object (either a file, or a `StringIO`). It should never return a string.

Reply:
What if it's constructed with a string, and `open` is set to `False`. Wouldn't that just return a string?

Reply:
It would, but it shouldn't be called with that mix, so I should guard against that in `FileWrapper.__init__`

Reply:
I have moved relevant logic into the `FileWrapper` to make it determine this state itself (and thereby avoid any issue where it can be improperly initialized).